lib/secretsscanner/authorizedkeys/authorized_keys.go

```diff
@@ -0,0 +1,391 @@
+	// maxReSendInterval is the maximum interval to re-send the authorized keys report
+	// to the cluster in case of no changes.
+	const maxReSendInterval = accessgraph.AuthorizedKeyDefaultKeyTTL - 20*time.Minute
+	timer := w.clock.NewTimer(jitterFunc(maxReSendInterval))
+	defer timer.Stop()
+	for {
+
+		if err := w.fetchAndReportAuthorizedKeys(ctx, stream, watcher); err != nil {
+			w.logger.Warn("Failed to report authorized keys", "error", err)
+		}
+
+		if !timer.Stop() {
+			<-timer.Chan()
+		}
+		timer.Reset(jitterFunc(maxReSendInterval))
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-reload:
+		case <-timer.Chan():
+		}
+	}
+}
+
+// isAuthorizedKeysReportEnabled checks if the cluster has authorized keys report enabled.
+func (w *Watcher) isAuthorizedKeysReportEnabled(ctx context.Context) (bool, error) {
+	accessGraphConfig, err := w.client.GetClusterAccessGraphConfig(ctx)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	return accessGraphConfig.GetEnabled() && accessGraphConfig.GetSecretsScanConfig().GetSshScanEnabled(), nil
+}
+
+// fetchAndReportAuthorizedKeys fetches the authorized keys from the system and reports them to the cluster.
+func (w *Watcher) fetchAndReportAuthorizedKeys(
+	ctx context.Context,
+	stream accessgraphsecretsv1pb.SecretsScannerService_ReportAuthorizedKeysClient,
+	watcher *fsnotify.Watcher,
+) error {
+	users, err := userList(ctx, w.logger, w.usersAccountFile)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var keys []*accessgraphsecretsv1pb.AuthorizedKey
+	for _, u := range users {
+		if u.HomeDir == "" {
+			w.logger.DebugContext(ctx, "Skipping user with empty home directory", "user", u.Name)
+			continue
+		}
+
+		authorizedKeysPath := filepath.Join(u.HomeDir, ".ssh", "authorized_keys")
+		if fs, err := os.Stat(authorizedKeysPath); err != nil || fs.IsDir() {
+			continue
+		}
+
+		hostKeys, err := w.parseAuthorizedKeysFile(u, authorizedKeysPath)
+		if err != nil {
+			w.logger.Warn("Failed to parse authorized_keys file", "error", err)
+			continue
+		}
+
+		// Add the file to the watcher. If file was already added, this is a no-op.
+		if err := watcher.Add(authorizedKeysPath); err != nil {
+			w.logger.Warn("Failed to add watcher for file", "error", err)
+		}
+		keys = append(keys, hostKeys...)
+	}
+
+	const maxKeysPerReport = 500
+	for i := 0; i < len(keys); i += maxKeysPerReport {
+		start := i
+		end := i + maxKeysPerReport
+		if end > len(keys) {
+			end = len(keys)
+		}
+		if err := stream.Send(
+			&accessgraphsecretsv1pb.ReportAuthorizedKeysRequest{
+				Keys:      keys[start:end],
+				Operation: accessgraphsecretsv1pb.OperationType_OPERATION_TYPE_ADD,
+			},
+		); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if err := stream.Send(
+		&accessgraphsecretsv1pb.ReportAuthorizedKeysRequest{Operation: accessgraphsecretsv1pb.OperationType_OPERATION_TYPE_SYNC},
+	); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// userList retrieves all users on the system
```

Comment:
This is half true, if you're using LDAP or something similar then users are not listed in a file AFAIR

Reply:
I do agree with this but to support it, I don't see any other way besides connecting to the ldap server, listing all users and extract their home dirs locations. 

Is there any other way?